ethcore/src/blockchain/blockchain.rs

```diff
@@ -1524,101 +1522,99 @@ mod tests {
-		assert_eq!(bc.best_block_hash(), first_hash.clone());
-		assert_eq!(bc.block_hash(1), Some(first_hash.clone()));
-		assert_eq!(bc.block_details(&first_hash).unwrap().parent, genesis_hash.clone());
-		assert_eq!(bc.block_details(&genesis_hash).unwrap().children, vec![first_hash.clone()]);
+		assert_eq!(bc.best_block_hash(), first_hash);
+		assert_eq!(bc.block_hash(1), Some(first_hash));
+		assert_eq!(bc.block_details(&first_hash).unwrap().parent, genesis_hash);
+		assert_eq!(bc.block_details(&genesis_hash).unwrap().children, vec![first_hash]);
 		assert_eq!(bc.block_hash(2), None);
 	}
 
 	#[test]
 	fn check_ancestry_iter() {
-		let mut canon_chain = ChainGenerator::default();
-		let mut finalizer = BlockFinalizer::default();
-		let genesis = canon_chain.generate(&mut finalizer).unwrap();
-		let genesis_hash = BlockView::new(&genesis).header_view().hash();
+		let genesis = BlockBuilder::genesis();
+		let first_10 = genesis.add_blocks(10);
+		let mut generator = BlockGenerator::new(vec![genesis, first_10]);
+
+		let genesis = generator.next().unwrap();
+		let genesis_hash = genesis.hash();
 
 		let db = new_db();
-		let bc = new_chain(&genesis, db.clone());
+		let bc = new_chain(&genesis.data(), db.clone());
 
-		let mut block_hashes = vec![genesis_hash.clone()];
+		let mut block_hashes = vec![genesis_hash];
 		let mut batch = db.transaction();
-		for _ in 0..10 {
-			let block = canon_chain.generate(&mut finalizer).unwrap();
-			block_hashes.push(BlockView::new(&block).header_view().hash());
-			bc.insert_block(&mut batch, &block, vec![]);
+		for block in generator {
+			block_hashes.push(block.hash());
+			bc.insert_block(&mut batch, &block.data(), vec![]);
 			bc.commit();
 		}
 		db.write(batch).unwrap();
 
 		block_hashes.reverse();
 
-		assert_eq!(bc.ancestry_iter(block_hashes[0].clone()).unwrap().collect::<Vec<_>>(), block_hashes)
+		assert_eq!(bc.ancestry_iter(block_hashes[0].clone()).unwrap().collect::<Vec<_>>(), block_hashes);
+		assert_eq!(block_hashes.len(), 11);
 	}
 
 	#[test]
 	fn test_find_uncles() {
-		let mut canon_chain = ChainGenerator::default();
-		let mut finalizer = BlockFinalizer::default();
-		let genesis = canon_chain.generate(&mut finalizer).unwrap();
-		let b1b = canon_chain.fork(1).generate(&mut finalizer.fork()).unwrap();
-		let b1a = canon_chain.generate(&mut finalizer).unwrap();
-		let b2b = canon_chain.fork(1).generate(&mut finalizer.fork()).unwrap();
-		let b2a = canon_chain.generate(&mut finalizer).unwrap();
-		let b3b = canon_chain.fork(1).generate(&mut finalizer.fork()).unwrap();
-		let b3a = canon_chain.generate(&mut finalizer).unwrap();
-		let b4b = canon_chain.fork(1).generate(&mut finalizer.fork()).unwrap();
-		let b4a = canon_chain.generate(&mut finalizer).unwrap();
-		let b5b = canon_chain.fork(1).generate(&mut finalizer.fork()).unwrap();
-		let b5a = canon_chain.generate(&mut finalizer).unwrap();
+		let genesis = BlockBuilder::genesis();
+		let b1a = genesis.add_block();
+		let b2a = b1a.add_block();
```

Comment:
Why not:
```
let chain_a = genesis.add_blocks(5);
let chain_b = genesis.add_blocks_with_difficulty(9);

let uncle_headers = chain_b[2..4].rev().map(|x| x.last_block_header()).collect();
```
or
```
let chain_a = genesis.add_blocks(5);
// with_difficulty clones the builder and changes the default options for all new generated blocks
let chain_b = genesis.with_difficulty(9).add_blocks(5);
```

Reply:
if you look closely, there are 5 different forks :)

```
genesis
|      \ 
b1a     b1b
|   \
|    \
|     \
|      \
b2a     b2b
|
....
```

Reply:
I didn't want to change the logic of a test